### ISSUE TITLE
Standard API for topk and / or argpartition

### ISSUE THREAD

**Issue by ogrisel** [data-apis/array-api#629]
numpy provides an indirect way to compute the smallest (or largest) values of an array [numpy.argpartition](https://numpy.org/doc/stable/reference/generated/numpy.argpartition.html) 

There is also a proposal to implement a simpler (arg)topk in numpy (without allocating a result array of the size of the input array when `k` is small):

- https://github.com/numpy/numpy/pull/19117

Here is a quick review of some available implementations in related libraries:

- [torch.topk](https://pytorch.org/docs/stable/generated/torch.topk.html) (no such thing as `torch.argpartition`)
   - returns a tuple of values and indices
- [jax.lax.top_k](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.top_k.html)
   - returns a tuple of values and indices
- [dask.array.topk](https://docs.dask.org/en/stable/generated/dask.array.topk.html)
   - returns only the values, I did not find a way to get the indices :(
- [cupy.argpartition](https://docs.cupy.dev/en/stable/reference/generated/cupy.argpartition.html) but internally computes a full `cupy.argsort` which makes it very inefficient  for large arrays and small `k`: O(nlog(n)) instead of O(n).

Motivation: (arg)topk is needed by popular baseline data-science workloads (e.g. k-nearest neighbors classification in scikit-learn) and is surprisingly non trivial to implement efficiently. For instance on GPUs, the fastest implementations are based on some kind of partial radix sort while CPU implementations would use more traditional partial sorting algorithms (as implemented in `std:partial_sort` or `std::nth_element`).

**Comment by ogrisel** [comment]
Note: since `argsort` is part of the standard Array API, it would be possible to implement a generic yet inefficient fallback in `array-api-compat` while allowing to dispatch to a more efficient routine for libraries that provide it.

**Comment by rgommers** [comment]
Thanks for the proposal @ogrisel. It's actually surprising that coverage and performance across array libraries is so spotty. I dug up the [NumPy mailing list discussion](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/F4P5UVTAKRJJ3OORI6UOWFSUEE5CNTSC/#PELUDW5ACUBHBNK5IVGWIWTQHBM2HXUP), and it seemed more or less positive, just unfinished and the name to use is a nicely-sized bikeshed.

Is this function something you already have in scikit-learn internally, or are you looking for something more efficient than the `argsort` or similar function you use now?

**Comment by ogrisel** [comment]
In scikit-learn, for k-nearest neighbors (bruteforce exact method for medium dim space), we use a routine optimized for multicore CPUs using Cython + OpenMP for pairwise distance (similar to scipy's `cdist`) fused with a topk reduction implemented in templated Cython. The topk reduction itself (called "argkmin" in scikit-learn) lives here:

- https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp

This code can only be called as reduction fused into the multithreaded pairwise distance computation. It is orchestrated via:

- https://github.com/scikit-learn/scikit-learn/blob/f5ec34e0f76277ba6d0a77d3033db0af83899b64/sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py#L157

For CPU, I doubt than any Array API based solution will be able to compete both on speed and memory usage.

However, we are interested in implementing Array API support for an alternative numpy code-path in order to provide GPU support, e.g. via PyTorch or CuPy. The reducer used in the numpy code-path is there:

- https://github.com/scikit-learn/scikit-learn/blob/f5ec34e0f76277ba6d0a77d3033db0af83899b64/sklearn/neighbors/_base.py#L704

It's based on `numpy.argpartition` followed by `numpy.argsort` of the top k values.

Note that to efficiently implement k-nearest neighbors in scikit-learn using the Array API, we would also need the Array API to provide [scipy.spatial.distance.cdist](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html).

I have not open an issue to discuss `cdist` yet. I wanted to probe the waters with `topk` first.

**Comment by shoyer** [comment]
JAX also has an approximate top-k implementation specifically tuned for TPUs: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.approx_max_k.html

**Comment by ogrisel** [comment]
I am not sure if we want to include non-exact methods in the spec. I have the feeling that there are many ways to compute such approximations and that they will require different and evolving parametrizations with different speed-accuracy trade-offs.

**Comment by kgryte** [comment]
A PR has now been opened which proposes adding `top_k` and friends to the specification: https://github.com/data-apis/array-api/pull/722. Please feel free to review and comment there with your concerns and feedback.

**Comment by jakirkham** [comment]
> [dask.array.topk](https://docs.dask.org/en/stable/generated/dask.array.topk.html)
returns only the values, I did not find a way to get the indices :(

There is [`dask.array.argtopk`]( https://docs.dask.org/en/stable/generated/dask.array.argtopk.html ), which simply returns the indices for the values `topk` would return

**Comment by kgryte** [comment]
@jakirkham Would be good to get your feedback on https://github.com/data-apis/array-api/pull/722.
